File: algorithmic-toolbox/src/week3/scala/MajorityElement.scala
package week3.scala

import java.util.Scanner

object MajorityElement {
  def main(args: Array[String]) = {
    val in = new Scanner(System.in)
    println(majorityElement((1 to in.nextInt).map(n => in.nextInt)))
  }

  def majorityElement(xs: Seq[Int]) = {
    if (xs.groupBy(_.toInt).mapValues(_.size).find(_._2 > xs.size/2).isDefined) 1
    else 0
  }
}

File: algorithmic-toolbox/src/week3/java/MajorityElement.java
package week3.java;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

class MajorityElement {

    public static void main(final String[] args) {
        try (final Scanner in = new Scanner(System.in)) {
            final Map<Integer, Integer> totals = new HashMap<>();
            final int count = in.nextInt();
            final int needed = ((count % 2) == 0) ? ((count/2) + 1) : ((count + 1) / 2);

            final Integer defaultValue = 0;
            for (int i = 0; i < count; i++) {
                final Integer key = in.nextInt();
                totals.put(key, totals.getOrDefault(key, defaultValue) + 1);
            }

            final boolean hasMajority = totals.values().stream().anyMatch(n -> n >= needed);
            System.out.println(hasMajority ? 1 : 0);
        }
    }
}